src/Orleans/Async/BatchWorker.cs

```diff
@@ -0,0 +1,175 @@
+﻿using Orleans;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Web;
+
+namespace Orleans
+{
+    /// <summary>
+    /// General pattern for an asynchronous worker that performs a work task, when notified,
+    /// to service queued work. Each work cycle handles ALL the queued work. 
+    /// If new work arrives during a work cycle, another cycle is scheduled. 
+    /// The worker never executes more than one instance of the work cycle at a time, 
+    /// and consumes no resources when idle. It uses TaskScheduler.Current 
+    /// to schedule the work cycles.
+    /// </summary>
+    public abstract class BatchWorker
+    {
+        // Subclass overrides this to define what constitutes a work cycle
+        protected abstract Task Work();
+
+        /// <summary>
+        /// Notify the worker that there is more work.
+        /// </summary>
+        public void Notify()
+        {
+            lock (this)
+            {
+                if (currentworkcycle != null)
+                {
+                    // lets the current work cycle know that there is more work
+                    morework = true;
+                }
+                else
+                {
+                    // start a work cycle
+                    Start();
+                }
+            }
+        }
+
+        // task for the current work cycle, or null if idle
+        private Task currentworkcycle;
+ 
+        // flag is set to indicate that more work has arrived during execution of the task
+        private volatile bool morework;
+
+        // is non-null if some task is waiting for the next work cycle to finish
+        private TaskCompletionSource<Task> nextworkcyclepromise;
+
+        private void Start()
+        {
+            // start the task that is doing the work
+            currentworkcycle = Work();
+
+            // chain a continuation that checks for more work, on the same scheduler
+            currentworkcycle.ContinueWith(t => this.CheckForMoreWork(), TaskScheduler.Current);
```

Comment:
Since this is an abstract utility class, it does not do logging.
Exception handling and logging, if wanted, must be done inside Work()